EIPS/eip-1283.md

```diff
@@ -0,0 +1,213 @@
+  the case if *current value* equals *new value*.
+* **Fresh**: this storage slot has not been changed, or has been reset
+  to its original value either on current frame, or on a sub-call
+  frame for the same contract. This is the case if *current value*
+  does not equal *new value*, and *original value* equals *current
+  value*.
+* **Dirty**: this storage slot has already been changed, either on
+  current frame or on a sub-call frame for the same contract. This is
+  the case if *current value* does not equal *new value*, and
+  *original value* does not equal *current value*.
+
+We can see that the above three types cover all possible variations of
+*original value*, *current value*, and *new value*.
+
+**No-op** is a trivial operation. Below we only consider cases for
+**Fresh** and **Dirty**.
+
+All initial (not-**No-op**) SSTORE on a particular storage slot starts
+with **Fresh**. After that, it will become **Dirty** if the value has
+been changed (either on current call frame or a sub-call frame for the
+same contract). When going from **Fresh** to **Dirty**, we charge the
+gas cost the same as current scheme.
+
+When entering a sub-call frame, a previously-marked **Dirty** storage
+slot will again become **Fresh**, but only for this sub-call
+frame. Note that we don't charge any more gases compared with current
+scheme in this case.
+
+In current call frame, a **Dirty** storage slot can be reset back to
+**Fresh** via a SSTORE opcode either on current call frame or a
+sub-call frame. For current call frame, this dirtiness is tracked, so
+we can issue refunds. For sub-call frame, it is not possible to track
+this dirtiness reset, so the refunds (for *current call frame*'s
+initial SSTORE from **Fresh** to **Dirty**) are not issued. In the
+case where refunds are not issued, the gas cost is the same as the
+current scheme.
+
+When a storage slot remains at **Dirty**, we charge 200 gas. In this
+case, we would also need to keep track of `R_SCLEAR` refunds -- if we
+already issued the refund but it no longer applies (*current value* is
+0), then removes this refund from the refund counter. If we didn't
+issue the refund but it applies now (*new value* is 0), then adds this
+refund to the refund counter. It is not possible where a refund is not
+issued but we remove the refund in the above case, because all storage
+slot starts with **Fresh** state, either on current call frame or a
+sub-call frame.
+
+### State Transition
+
+Below is a graph ([by
+@Arachnid](https://github.com/ethereum/EIPs/pull/1283#issuecomment-410229053))
+showing possible state transition of gas costs. Note that this applies
+to current call frame only, and we ignore **No-op** state because that
+is trivial:
+
+![State Transition](../assets/eip-1283/state.png)
+
+Below are table version of the above diagram. Horizontal shows the
+next state, and vertical shows the current state.
+
+|                  | `current=orig=0`         | `current!=orig`    |
```

Comment:
Wouldn't this be clearer if one axis had `(current, original)` and the other axis had `new`?